addOns/authhelper/src/main/java/org/zaproxy/addon/authhelper/AuthenticationDetectionScanRule.java

```diff
@@ -0,0 +1,268 @@
+                new AuthenticationRequestDetails(
+                        null,
+                        new HtmlParameter(HtmlParameter.Type.form, "username", "test"),
+                        new HtmlParameter(HtmlParameter.Type.form, "password", "test"),
+                        AuthDataType.FORM,
+                        null,
+                        Collections.emptyList(),
+                        0);
+        alerts.add(this.getAlert(authDetails).build());
+        return alerts;
+    }
+
+    protected HtmlParameter getMatchingParam(
+            TreeSet<HtmlParameter> params, String[] fullNames, String[] elements) {
+        // First try to get an exact match
+        Optional<HtmlParameter> first =
+                params.stream()
+                        .filter(
+                                i ->
+                                        Arrays.stream(fullNames)
+                                                .anyMatch(p -> i.getName().equalsIgnoreCase(p)))
+                        .findFirst();
+        if (first.isPresent()) {
+            return first.get();
+        }
+        // Next try the elements
+        first =
+                params.stream()
+                        .filter(
+                                i ->
+                                        Arrays.stream(elements)
+                                                .anyMatch(
+                                                        p -> i.getName().toLowerCase().contains(p)))
+                        .findFirst();
+        if (first.isPresent()) {
+            return first.get();
+        }
+        return null;
+    }
+
+    private void extractJsonStrings(
+            JSONObject jsonObject, String parent, TreeSet<HtmlParameter> params) {
+        for (Object key : jsonObject.keySet()) {
+            Object obj = jsonObject.get(key);
+            if (obj instanceof JSONObject) {
+                extractJsonStrings((JSONObject) obj, normalisedKey(parent, (String) key), params);
+            } else if (obj instanceof String) {
+                params.add(
+                        new HtmlParameter(
+                                HtmlParameter.Type.form,
+                                normalisedKey(parent, (String) key),
+                                (String) obj));
+            }
+        }
+    }
+
+    private String normalisedKey(String parent, String key) {
+        return parent.isEmpty() ? key : parent + "." + key;
+    }
+
+    @Override
+    public String getName() {
+        return Constant.messages.getString("authhelper.auth-detect.name");
+    }
+
+    public String getDescription() {
+        return Constant.messages.getString("authhelper.auth-detect.desc");
+    }
+
+    public String getSolution() {
+        return Constant.messages.getString("authhelper.auth-detect.soln");
+    }
+
+    public String getReference() {
+        return "https://www.zaproxy.org/docs/desktop/addons/authhelper/";
```

Comment:
It would be preferable to get it from the `AddOn` class, which should be exposed through the extension. Otherwise it needs to be updated per comment in the manifest.